server/src/main/java/com/linecorp/centraldogma/server/internal/api/ContentServiceV1.java

```diff
@@ -281,73 +274,53 @@ private Object handleWatchFailure(Throwable thrown) {
     }
 
     /**
-     * DELETE /projects/{projectName}/repos/{repoName}/contents{path}?revision={revision}
+     * GET /projects/{projectName}/repos/{repoName}/commits/{revision}?path={path}&to={to}
      *
-     * <p>Deletes a file.
+     * <p>Returns the list of commits in the path.
      */
-    @Delete("regex:/projects/(?<projectName>[^/]+)/repos/(?<repoName>[^/]+)/contents(?<path>(|/.*))$")
-    public CompletionStage<Void> deleteFile(@Param("path") String path,
-                                            @Param("revision") @Default("-1") String revision,
-                                            @RequestObject Repository repository,
-                                            @RequestObject JsonNode node,
-                                            @RequestObject Author author) {
-        final String path0 = rootDirIfEmpty(path);
-
-        checkArgument(node.get("commitMessage") != null, "commit message should be non-null");
-        final JsonNode commitMessageNode = node.get("commitMessage");
-
-        final CommitMessageDto commitMessage = convertCommitMessage(commitMessageNode);
-        final long commitTimeMillis = System.currentTimeMillis();
+    @Get("regex:/projects/(?<projectName>[^/]+)/repos/(?<repoName>[^/]+)/commits(?<revision>(|/.*))$")
+    public CompletionStage<?> listCommits(@Param("revision") String revision,
+                                          @Param("path") @Default("/**") String path,
+                                          @Param("to") Optional<String> to,
+                                          @RequestObject Repository repository) {
+        final Revision fromRevision;
+        final Revision toRevision;
+
+        // 1. only the "revision" is specified:       get the "revision"
+        // 2. only the "to" is specified:             get from "HEAD" to "to"
+        // 3. the "revision" and "to" is specified:   get from the "revision" to "to"
+        // 4. nothing is specified:                   get from "HEAD" to "INIT"
```

Comment:
The case 2 and 4 would beg for memory spike or OOME with the current `Repository` API, which does not stream commits. Could we limit the number of commits retrieved?

Reply:
So just return the first `N` commits where `N` is configurable?
Also, I feel strongly that we need to implement streaming commits using `Relation` since we are going to return just the part of the request.

Reply:
Yes and yes.